Update CI docu (#846)
ElektraInitiative/PermaplanT: doc/ci.md

## Triggers

> The pipeline will be executed ONLY for pushes to Pull Requests.

Reviewer: ```suggestion
The pipeline will be executed **only** for pushes to Pull Requests and for pushes to master.
```

> If you want the pipeline to be executed, you can add a comment saying `jenkins build please`.

Reviewer: Better describe when/where this should be done. (In PRs, if there are spurious problems)

If problems persist, please create a new issue with the failing build log.

For users with login credentials for Jenkins, you can manually execute the pipeline for a branch or pull request via the [Jenkins UI](https://build.libelektra.org/job/PermaplanT/).
Please ask if you need login data.

## Concurrent Builds

Concurrent builds on the same branch are disabled.
The master branch is excluded from this rule.

## Stages

### Sanity stage

This stage performs rapid checks like pre-commit, migrations and schema building.

Local pre-commit currently does not perform the codespell hook, but it is performed in the pipeline.
Cargo fmt, eslint and groovy linting is not performed through pre-commit.

### Tests and Build

This is a parallel stage which fails fast (exits if one stage fails) or times out after 2 hours.

It can be subdivided in 3 categories:

- Parallel cargo stages
- Sequential frontend stages
- Sequential mdbook stages

> Following `hidden` checks are done:

Reviewer: What do you mean with `hidden`?

Author: Some tests that you would not guess that they are there. This is actually an issue and should be made more transparent.

### Tests and Build

This is a parallel stage that has a 2 hour timeout and exits if one substage fails.
It checkouts the current codebase and each stage runs inside a separate docker container.
This is a parallel stage which fails fast (exits if one stage fails) or times out after 2 hours.

Following substages exist:
It can be subdivided in 3 categories:

- Test and Build Mdbook
- Test and Build Backend (With PostGIS sidecar container)
- Test and Build Frontend
- Parallel cargo stages
- Sequential frontend stages
- Sequential mdbook stages

For more information about automated integration tests look [here](./tests/README.md).
Following `hidden` checks are done:

> Every pipeline run has its own isolated database.

Reviewer: Is this not true anymore?

Author: It is true, but I removed some information that I think is obvious or not important for the reader.

> Jenkins will acquires a lock here and releases it after finishing the E2E tests.

Reviewer: ```suggestion
Jenkins will acquire a lock during execution of E2E tests.
```

> E2E tests are run on [dev.permaplant.net](https://dev.permaplant.net).

Reviewer: Isn't it pr only?

```suggestion
E2E tests are only executed on [pr.permaplant.net](https://pr.permaplant.net).
```

Author: Yes they get performed only on pr.permaplant.net.
Maybe we would want to run them on dev.permaplant.net when merged to master?

Author: Oh I just realized thats a type, I meant pr.permaplant

> Test reports and results are generated after successful execution.

Reviewer: Please describe how to find them when coming from "Details" of GitHub, i.e., e.g., for 846 https://build.libelektra.org/blue/organizations/jenkins/PermaplanT/detail/PR-846/5/pipeline

Author: Is it well enough and clearly described now?

> - A [html report](https://build.libelektra.org/blue/organizations/jenkins/PermaplanT/detail/master/395/artifacts)

Reviewer: Why 395? Maybe also point to last completed build?

- Videos and screenshots can on failed tests can also be found where the html report is

### Deploy Dev

The `master` branch will be automatically deployed to [dev.permaplant.net](https://dev.permaplant.net).

> ### Deploy Prod

Reviewer: Actually, here should be only written that the release pipeline is only relevant for doc/release.md (if this section is needed at all)

It can be subdivided in 3 categories:

#### Multiple parallel cargo stages

Standard cargo build, clippy, doc and test is performed here.

#### One sequential frontend stages

Frontend is built, tested, linted, format checked and additionally package.json version is verfied to be up to date.
Storybook and typedocs are also generated here.

#### One sequential mdbook stages

Mdbook is built, tested and links are checked.

### Deploy PR

Every pull request will be deployed on a publicly available instance on [pr.permaplant.net](https://pr.permaplant.net).
Jenkins will acquires a lock before deploying to pr.permaplant.net and release it after finishing the E2E stage.
This lock prevents other jobs to overwrite the deployment while we are performing E2E tests.

Since there is only one agent for PRs available, the last built PR wins.

### E2E Tests

E2E tests are run on [dev.permaplant.net](https://dev.permaplant.net).
If a test times out, it is retried.

[Test reports and results](https://build.libelektra.org/job/PermaplanT/job/master/lastCompletedBuild/artifact/e2e/) are generated on successful execution.

> They contain videos, screenshots and a html report.

Author: That test just timed out and didnt even connect to the webapp.

## Triggers

> The pipeline will be executed **only** for pushes to Pull Requests and for pushes to master.

Reviewer: I think he would like to have a clarification that **all** pushes to Pull Requests get built, maybe something like:

```suggestion
The pipeline will be executed for:
- all pushes to branches for which a Pull Requests exists, starting with the creation of the pull request.
- all pushes to master.
```

@tschawax does my text clarify what you wanted to know?

If the pipeline fails and you think it is not the fault of your code, you can re-execute the pipeline by typing `jenkins build please` as a comment.
If problems persist, please create a new issue with the failing build log.

For users with login credentials for Jenkins, you can manually execute the pipeline for a branch or pull request via the [Jenkins UI](https://build.libelektra.org/job/PermaplanT/).
Please ask if you need login data.

## Concurrent Builds

> Concurrent builds on the same branch are disabled.

Author: Only the newest runs.
Yes manual if needed and you have the permissions.

> Concurrent builds on the same branch are disabled.

Reviewer: I answered in #874.
